File: leaderboard/personal/go/parse.go
package main

import (
	"bytes"
	"errors"
	"fmt"
	"golang.org/x/net/html"
	"io"
	"os"
	"strconv"
	"strings"
	"time"
)

func extractMainHtml(doc *html.Node) (*html.Node, error) {
	var body *html.Node
	var crawler func(*html.Node)
	crawler = func(node *html.Node) {
		if node.Type == html.ElementNode && node.Data == "main" {
			body = node
			return
		}
		for child := node.FirstChild; child != nil; child = child.NextSibling {
			crawler(child)
		}
	}
	crawler(doc)
	if body != nil {
		return body, nil
	}
	return nil, errors.New("Missing <main> in the node tree")
}

func renderNode(n *html.Node) string {
	var buf bytes.Buffer
	w := io.Writer(&buf)
	html.Render(w, n)
	return buf.String()
}

func parseMaybeInt(s string) IntStat {
	switch s {
	case "-":
		return IntStat{
			Type:  MissingInt,
			Value: 0,
		}
	default:
		n, err := strconv.Atoi(s)
		if err != nil {
			fmt.Println("[ERROR] Failed to parse integer: ", err)
		}
		return IntStat{
			Type:  ValidInt,
			Value: n,
		}
	}
}

func parseMaybeDuration(s string) DurationStat {
	switch s {
	case "-":
		return DurationStat{
			Type:     MissingDuration,
			Duration: time.Duration(0) * time.Second,
		}
	case ">24h":
		return DurationStat{
			Type:     OverADayDuration,
			Duration: time.Duration(24)*time.Hour + time.Duration(1)*time.Second,
		}
	default:
		t1 := time.Date(0, time.January, 1, 0, 0, 0, 0, time.UTC)
		t2, err := time.Parse("15:04:05", s)
		if err != nil {
			fmt.Println("[ERROR] Failed to parse time from string: ", err)
			os.Exit(1)
		}

		duration := t2.Sub(t1)
		return DurationStat{
			Type:     PreciseDuration,
			Duration: duration,
		}
	}
}

func ExtractStatsHtml(doc *html.Node) (map[int]DayStats, error) {
	statsHtml, err := extractMainHtml(doc)
	if err != nil {
		fmt.Println("[ERROR] Failed to extract main stats from HTML: ", err)
		os.Exit(1)
	}

	// We are going to parse the stats from string values because
	// there's no nice way to get the results otherwise.  The stats
	// is not in table format.

	// Render the main node as a string and split it into lines
	statsHtmlStr := html.UnescapeString(renderNode(statsHtml))
	statsStrLines := strings.Split(statsHtmlStr, "\n")
	statsStrLines = statsStrLines[3 : len(statsStrLines)-3]

	// Initialise the output map
	stats := make(map[int]DayStats)

	// Parse stats from each line
	for i := 0; i < len(statsStrLines); i++ {
		line := statsStrLines[i]
		words := strings.Fields(line)

		// Parse integer fields
		year := IntStat{
			Type:  ValidInt,
			Value: AOC_YEAR,
		}
		day := parseMaybeInt(words[0])
		rank1 := parseMaybeInt(words[2])
		score1 := parseMaybeInt(words[3])
		rank2 := parseMaybeInt(words[5])
		score2 := parseMaybeInt(words[6])

		// Parse time fields
		time1 := parseMaybeDuration(words[1])
		time2 := parseMaybeDuration(words[4])

		// Make day stats and push to map
		var parts [2]DayPartStats
		parts[0] = DayPartStats{
			Year:  year,
			Day:   day,
			Time:  time1,
			Rank:  rank1,
			Score: score1,
		}
		parts[1] = DayPartStats{
			Year:  year,
			Day:   day,
			Time:  time2,
			Rank:  rank2,
			Score: score2,
		}
		stats[day.Value] = DayStats{
			Year:  year,
			Day:   day,
			Parts: parts,
		}
	}
	return stats, nil
}

File: leaderboard/personal/rs/src/parse.rs
use super::{
    stats::{Date, DayPartStats, DayStats, Duration, Integer},
    AOC_YEAR,
};
use chrono::{self, NaiveTime};
use scraper::{Html, Selector};
use std::collections::HashMap;

// https://github.com/causal-agent/scraper
// Uses html5 under the hood
pub fn extract_stats(html: String) -> HashMap<usize, DayStats> {
    let document = Html::parse_document(&html);
    let main_selector = Selector::parse("main").unwrap();
    let article_selector = Selector::parse("article").unwrap();
    let pre_selector = Selector::parse("pre").unwrap();

    let main = document.select(&main_selector).next().unwrap();
    let article = main.select(&article_selector).next().unwrap();
    let pre = article.select(&pre_selector).next().unwrap();

    extract_stats_from_str(pre.inner_html())
}

fn extract_stats_from_str(stats: String) -> HashMap<usize, DayStats> {
    let mut map: HashMap<usize, DayStats> = HashMap::new();

    // Skip the preamble (headers) that we don't care about
    for line in stats.lines().skip(2) {
        // TODO: consider using something like:
        //   - github.com/veddan/rust-htmlescape; or
        //   - github.com/phaazon/html-entities.
        let line = line.replace("&gt;", ">");
        let words: Vec<_> = line.split_whitespace().collect();

        // Parse integers
        let day = parse_int_from_stats(words[0]).unwrap(); // We should always have a day
        let rank1 = parse_int_from_stats(words[2]);
        let score1 = parse_int_from_stats(words[3]);
        let rank2 = parse_int_from_stats(words[5]);
        let score2 = parse_int_from_stats(words[6]);

        // Parse durations
        let time1 = parse_duration_from_stats(words[1]);
        let time2 = parse_duration_from_stats(words[4]);

        // Make day stats and push to map
        let parts = DayStats {
            date: Date {
                day,
                year: AOC_YEAR,
            },
            parts: [
                DayPartStats {
                    date: Date {
                        day,
                        year: AOC_YEAR,
                    },
                    part: 1,
                    time: time1,
                    rank: rank1,
                    score: score1,
                },
                DayPartStats {
                    date: Date {
                        day,
                        year: AOC_YEAR,
                    },
                    part: 2,
                    time: time2,
                    rank: rank2,
                    score: score2,
                },
            ],
        };
        map.insert(day, parts);
    }
    map
}

fn parse_int_from_stats(n: &str) -> Integer {
    match n {
        "-" => Integer::Missing,
        _ => match n.parse::<usize>() {
            Ok(n) => Integer::Some(n),
            Err(err) => {
                println!("[WARNING] Failed to parse integer {n:?} as type `usize': {err}");
                Integer::Unknown
            }
        },
    }
}

fn parse_duration_from_stats(d: &str) -> Duration {
    match d {
        "-" => Duration::Missing,
        ">24h" => Duration::OverOneDay,
        _ => {
            let t1 = NaiveTime::parse_from_str(d, "%H:%M:%S");
            match t1 {
                Ok(t1) => {
                    let t2 = NaiveTime::from_hms_opt(0, 0, 0).unwrap();
                    let duration = t1.signed_duration_since(t2);
                    Duration::Some(duration)
                }
                Err(err) => {
                    println!("[WARNING] Failed to parse time string {d:?} as type `chrono::Duration': {err}");
                    Duration::Unknown
                }
            }
        }
    }
}

impl Integer {
    fn unwrap(&self) -> usize {
        match self {
            Integer::Some(n) => *n,
            _ => panic!("Could not unwrap custom `Integer' type {:?}", self),
        }
    }
}
